core/container.go

```diff
@@ -687,6 +688,58 @@ func (container *Container) WithNewFile(ctx context.Context, dest string, conten
 	})
 }
 
+func (container *Container) WithSymlink(ctx context.Context, srv *dagql.Server, target, linkName string) (*Container, error) {
+	container = container.Clone()
+
+	op, ok := DagOpFromContext[ContainerDagOp](ctx)
+	if !ok {
+		return nil, fmt.Errorf("no dagop")
+	}
+
+	containerPath := path.Clean(path.Join(container.Config.WorkingDir, linkName))
+	linkNameDirPath, _ := filepath.Split(containerPath)
+
+	dir, mount, err := locatePath(container, linkNameDirPath, NewDirectory)
+	if err != nil {
+		return nil, err
+	}
+	if mount != nil {
+		// symlink will be added to a mounted directory
+		newDir, err := dir.WithSymlink(ctx, srv, target, strings.TrimPrefix(containerPath, mount.Target+"/"))
+		if err != nil {
+			return nil, err
+		}
+		mount.Result = newDir.Result
+		return container, nil
+	}
+
+	// otherwise symlink will be added to root fs
```

Comment:
:thinking: could we use the directory logic from above as well here? Maybe using `RootFS` + `WithRootFS` or similar?

Just a thought, but it would be *nice* if the logic here could be shared between here and `directory.go` - obviously if it's not particularly easy, not worth forcing it, but given we want to convert *every* operation to use dagop like this, then we'll see this pattern a lot :tada:

Reply:
just curious if you saw the existing `writeToPath` method: https://github.com/sipsma/dagger/blob/84dc5b3ba37238d957db1025c70dca797d35a055/core/container.go#L1099-L1099

seems like exactly what we'd want here, though it does internally do a `WithRootFS` call

---

> which does show up, I'm left wondering if WithRootFS needs some extra attention to make it work with dagops?

Yeah, looks like it. Probably just needs to check if the dir provided as a rootfs has a `.Result` set and, if so, set that to the container's `FSResult`

Reply:
I added check for `.Result` under alexcb@cae09e8bedd07b9a8297d89af4e2db5908ab2309, and it seems to be working now :tada: 

I initially went looking into using `writeToPath` when I first started on this PR, but ran into difficulties using it -- possibly due to the missing `.Result` check? I tried using it again, but tests started to fail, so I'll need to look into this deeper.

Reply:
`writeToPath` has been changed to use a new function `ContainerMounts.Replace` which prevents the mount index ordering from being changed.